api/src/main/java/com/epam/pipeline/manager/datastorage/providers/gcp/GSBucketStorageHelper.java

```diff
@@ -378,6 +381,72 @@ public void restoreFileVersion(final GSBucketStorage storage, final String path,
+    }
+
+    private void cleanDeleteMarkers(final Storage client,
+                                    final String bucketName, final String requestPath,
+                                    final RestoreFolderVO restoreFolderVO) {
+        String folderPath = Optional.ofNullable(requestPath).orElse(EMPTY_PREFIX);
+        if (StringUtils.isNotBlank(folderPath)) {
+            folderPath = normalizeFolderPath(requestPath);
+        }
+        final Page<Blob> blobs = client.list(bucketName,
+                Storage.BlobListOption.versions(true),
+                Storage.BlobListOption.currentDirectory(),
+                Storage.BlobListOption.prefix(folderPath),
+                Storage.BlobListOption.pageToken(EMPTY_PREFIX),
+                Storage.BlobListOption.pageSize(Integer.MAX_VALUE));
+        Assert.isTrue(Objects.nonNull(blobs) && blobs.iterateAll().iterator().hasNext(), messageHelper
+                .getMessage(MessageConstants.ERROR_DATASTORAGE_PATH_NOT_FOUND, folderPath, bucketName));
+        listItemsWithVersions(blobs)
+                .forEach(item -> {
+                    recursiveRestoreFolderCall(item, client, bucketName, restoreFolderVO);
+                    if (isFileWithDeleteMarkerAndShouldBeRestore(item, restoreFolderVO)) {
+                        final Blob blob = checkBlobExistsAndGet(bucketName, item.getPath(), client,
+                                removeDeletedMarkerFromVersion(((DataStorageFile) item).getVersion()));
+                        final Storage.CopyRequest request = Storage.CopyRequest.newBuilder()
+                                .setSource(blob.getBlobId())
+                                .setSourceOptions(Storage.BlobSourceOption.generationMatch())
+                                .setTarget(BlobId.of(bucketName, item.getPath()))
+                                .build();
+                        client.copy(request).getResult();
+                    }
+                });
+    }
+
+    private String removeDeletedMarkerFromVersion(final String versionWithDeletedMarker) {
+        if (latestVersionHasDeletedMarker(versionWithDeletedMarker)) {
+            return versionWithDeletedMarker.substring(0, versionWithDeletedMarker.length() - 2);
+        }
+        throw new DataStorageException(
+                String.format("Corresponded version: '%s' should has deleted marker: '%s'", versionWithDeletedMarker,
+                        LATEST_VERSION_DELETION_MARKER));
+    }
+
+    private void recursiveRestoreFolderCall(final AbstractDataStorageItem item, final Storage client,
+                                            final String bucketName, final RestoreFolderVO restoreFolderVO) {
+        if (item.getType() == DataStorageItemType.Folder && restoreFolderVO.isRecursively()) {
+            cleanDeleteMarkers(client, bucketName, item.getPath(), restoreFolderVO);
+        }
+    }
+
+    private boolean isFileWithDeleteMarkerAndShouldBeRestore(final AbstractDataStorageItem item,
```

Comment:
Maybe we should move this method into `ProviderUtils` class and reuse it into `S3Helper`?

Reply:
duplicate)